File: macos/Sources/VoiceTyper/Support/Localization/L10n.swift
import Foundation

/// 界面语言。中文是默认语言，英文为对等的第二语言。
///
/// 刻意**不做**「跟随系统」：识别语言（`ASRLanguage`）已经是一个独立的语言选项，
/// 再加一个会随系统漂移的界面语言，用户很难判断"我现在看到的是哪一种、为什么变了"。
/// 显式两选一，读配置文件也一眼看得懂。
enum AppLanguage: String, Codable, CaseIterable, Sendable {
    case zh
    case en

    /// 语言名用它自己的语言书写（与系统语言选择器的惯例一致），不随当前界面语言变化。
    var displayName: String {
        switch self {
        case .zh: return "中文"
        case .en: return "English"
        }
    }
}

/// 运行期界面语言与中英文案表。
///
/// **键就是中文原文**：`L("录音中")`。这样做的理由是
/// 1. 中文是默认语言，查不到翻译时回落到键本身仍然是正确的中文界面，不会出现 `hud.recording`
///    这种键名泄漏到界面上；
/// 2. 新增文案只需在一处补英文，不必同时维护一张键名表；
/// 3. 覆盖率可测：`LocalizationTests` 扫描全部源码里的 `L(...)` / `LF(...)`，要求每条都能在
///    英文表里查到，漏翻会让测试失败，而不是等用户切到英文才发现。
///
/// 语言在进程启动早期由 `bootstrap()` 定下，运行期改设置需要重启才生效（设置页有明确提示）。
enum L10n {
    private static let lock = NSLock()
    /// `nonisolated(unsafe)` + 锁：文案会在识别、下载、LLM 等后台线程被读取，
    /// 不能假定只在主线程访问。
    nonisolated(unsafe) private static var storedLanguage: AppLanguage = .zh

    /// 界面语言在 `UserDefaults` 里的镜像。
    ///
    /// `config.yaml` 仍是唯一的事实来源，但菜单栏与主菜单在配置读出来之前就已经构造完成，
    /// 那时再改语言就来不及了。镜像值只用于启动瞬间取一个语言，每次加载配置都会被同步。
    static let defaultsKey = "ui.interfaceLanguage"

    static var current: AppLanguage {
        lock.lock()
        defer { lock.unlock() }
        return storedLanguage
    }

    /// 进程启动早期调用：从 `UserDefaults` 镜像取语言。
    static func bootstrap(defaults: UserDefaults = .standard) {
        guard let raw = defaults.string(forKey: defaultsKey), let language = AppLanguage(rawValue: raw) else {
            return
        }
        setLanguage(language)
    }

    /// 配置加载/保存后调用：以 `config.yaml` 为准，并同步镜像供下次启动使用。
    static func apply(_ language: AppLanguage, defaults: UserDefaults = .standard) {
        setLanguage(language)
        defaults.set(language.rawValue, forKey: defaultsKey)
    }

    static func setLanguage(_ language: AppLanguage) {
        lock.lock()
        storedLanguage = language
        lock.unlock()
    }

    /// 查表。查不到（尚未翻译）时回落到中文原文，界面不会因此出现空白或键名。
    static func string(_ zh: String) -> String {
        switch current {
        case .zh:
            return zh
        case .en:
            return englishTable[zh] ?? zh
        }
    }
}

/// 取一条本地化文案。参数就是中文原文。
func L(_ zh: String) -> String {
    L10n.string(zh)
}

/// 取一条带占位符的本地化文案。中文原文即格式串，例如 `LF("下载模型 %d%%", percent)`。
func LF(_ zhFormat: String, _ arguments: CVarArg...) -> String {
    String(format: L10n.string(zhFormat), arguments: arguments)
}

File: windows/Support/Localization/L10n.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text.RegularExpressions;

namespace VoiceTyper.Support;

/// <summary>
/// 界面语言。中文是默认语言，英文为对等的第二语言（与 macOS <c>AppLanguage</c> 同构）。
/// </summary>
internal enum AppLanguage
{
    Zh,
    En,
}

internal static class AppLanguageExtensions
{
    public static string ToYamlValue(this AppLanguage language) => language switch
    {
        AppLanguage.En => "en",
        _ => "zh",
    };

    /// <summary>语言名用它自己的语言书写，不随当前界面语言变化。</summary>
    public static string DisplayName(this AppLanguage language) => language switch
    {
        AppLanguage.En => "English",
        _ => "中文",
    };

    public static AppLanguage Parse(string? value) => (value ?? "zh").Trim().ToLowerInvariant() switch
    {
        "en" => AppLanguage.En,
        _ => AppLanguage.Zh,
    };
}

/// <summary>
/// 运行期界面语言与中英文案表（与 macOS <c>L10n</c> 同构，键即中文原文）。
///
/// 键用中文原文而不是 <c>hud.recording</c> 这样的符号名，理由有三：中文是默认语言，
/// 查不到翻译时回落到键本身仍是正确的中文界面；新增文案只需补一处英文；覆盖率可测
/// —— <c>LocalizationTests</c> 扫描源码里的 <c>T(...)</c> / <c>F(...)</c>，漏翻直接让测试失败。
///
/// 语言在进程启动早期由 <see cref="Bootstrap"/> 定下，运行期改设置需要重启才全面生效
/// （设置页有明确提示）。
/// </summary>
internal static partial class L10n
{
    private static readonly object _gate = new();
    private static AppLanguage _current = AppLanguage.Zh;

    public static AppLanguage Current
    {
        get { lock (_gate) { return _current; } }
    }

    public static void SetLanguage(AppLanguage language)
    {
        lock (_gate) { _current = language; }
    }

    /// <summary>
    /// 进程启动早期调用：托盘菜单与设置窗口都是带着文案构造出来的，语言必须在那之前定下。
    ///
    /// 这里刻意不走 <c>ConfigStore.LoadOrCreate()</c>：那条路径会创建目录、写默认配置、
    /// 触发老配置迁移，只为读一个字段而在启动最早期产生这些副作用并不合适。读不到、
    /// 读出无法识别的值，一律回落中文。
    /// </summary>
    public static void Bootstrap(string? configPath = null)
    {
        try
        {
            var path = configPath ?? AppConstants.ConfigFilePath;
            if (!File.Exists(path)) return;

            var match = Regex.Match(
                File.ReadAllText(path),
                @"^\s*interface_language\s*:\s*[""']?(?<value>[A-Za-z_-]+)[""']?\s*$",
                RegexOptions.Multiline
            );
            if (match.Success)
            {
                SetLanguage(AppLanguageExtensions.Parse(match.Groups["value"].Value));
            }
        }
        catch (Exception ex)
        {
            // 语言读取失败绝不该阻断启动：回落中文，留一条记录即可。
            Console.Error.WriteLine($"[VoiceTyper] 读取界面语言失败，回落中文: {ex.Message}");
        }
    }

    /// <summary>配置加载/保存后调用，以 config.yaml 为准。</summary>
    public static void Apply(AppLanguage language) => SetLanguage(language);

    /// <summary>查表。查不到（尚未翻译）时回落到中文原文，界面不会出现空白或键名。</summary>
    public static string T(string zh)
    {
        if (Current == AppLanguage.Zh) return zh;
        return EnglishTable.TryGetValue(zh, out var translated) ? translated : zh;
    }

    /// <summary>带占位符的文案。中文原文即格式串，例如 <c>F("下载模型 {0}%", percent)</c>。</summary>
    public static string F(string zh, params object?[] arguments) =>
        string.Format(CultureInfo.CurrentCulture, T(zh), arguments);

    internal static IReadOnlyDictionary<string, string> Table => EnglishTable;
}
